File: #4 The box problem/Codes/Box.py
"""
CSU0018 Computer Algorithms
Programming Assignment 04
The Box Problem
"""
class box:
    def __init__(self, side_lst):
        side_lst.sort(reverse = True)
        self.side1, self.side2, self.side3 = side_lst
        self.vol = self.side1 * self.side2 * self.side3

    def can_contain(self, box):
        return self.side1 > box.side1 and self.side2 > box.side2 and self.side3 > box.side3

def main(boxes):
    boxes.sort(key = lambda box : box.vol)      # O(nlogn)
    adj_lst = {boxes[i] : [box for box in boxes[i:] if box.can_contain(boxes[i])] for i in range(len(boxes))}

    depth_table = {box : 1 for box in boxes}    # O(n^2)
    for box in boxes:
        for adj in adj_lst[box]:
            if depth_table[adj] < depth_table[box] + 1:
                depth_table[adj] = depth_table[box] + 1
    print(max(depth_table.values()))
#####################################################################################
SAMPLE_NO = 3
if __name__ == "__main__":
    main([box([int(line[0]), int(line[1]), int(line[2])]) for line in list(lines.strip('\n').split() for lines in open("test" + str(SAMPLE_NO) + ".in", 'r'))[1:]])

File: #4 The box problem/Codes/Box.cpp
/*
 * CSU0018 Computer Algorithms
 * Programming Assignment 04
 * The Box Problem
 */
#include <iostream>
#include <unordered_map>
#include <vector>
#include <algorithm>

class Box {
public:
	Box(int H, int W, int D, int i = 0) {
		std::vector<int> sides = {H, W, D};
		std::sort(sides.begin(), sides.end());
		_H = sides[0];
		_W = sides[1];
		_D = sides[2];
		vol = _H * _W * _D;
		NO = i;
	}
	bool can_contain(Box box) { return _H > box._H && _W > box._W && _D > box._D; }
	int vol;
	int NO;
private:
	int _H;
	int _W;
	int _D;
};

void max_boxing(std::vector<Box> &);

int main(void) {
	int lines;
	while(std::cin >> lines) {
		std::vector<Box> boxes;
		for(int i = 0; i < lines; ++i) {
			int h, w, d;
			std::cin >> h >> w >> d;
			boxes.push_back(Box(h, w, d, i));
		}
		max_boxing(boxes);
	}
	return 0;
}

void max_boxing(std::vector<Box> &boxes) {
	std::sort(boxes.begin(), boxes.end(), [&](Box a, Box b){ return a.vol < b.vol; });

	std::unordered_map<int, std::vector<Box>> adj_lst;
	for(size_t i = 0; i < boxes.size(); ++i) {
		std::vector<Box> adj;
		for(auto iter = boxes.begin() + i; iter != boxes.end(); ++iter)
			if((*iter).can_contain(boxes[i]))
				adj.push_back(*iter);
		adj_lst[boxes[i].NO] = adj;
	}

	std::unordered_map<int, int> depth_table;
	for(auto &box : boxes)
		depth_table[box.NO] = 1;

	int max_depth = 0;
	for(auto &box : boxes) {
		for(auto &adj : adj_lst[box.NO]) {
			int depth = depth_table[box.NO] + 1;
			if(max_depth < depth)
				max_depth = depth;
			if(depth_table[adj.NO] < depth)
				depth_table[adj.NO] = depth;
		}
	}

	std::cout << max_depth;
}